swarm/storage/ldbstore.go

```diff
@@ -549,26 +551,44 @@ func (s *LDBStore) doPut(chunk *Chunk, index *dpaDBIndex, po uint8) {
 }
 
 func (s *LDBStore) writeBatches() {
-	for range s.batchesC {
-		s.lock.Lock()
-		b := s.batch
-		e := s.entryCnt
-		d := s.dataIdx
-		a := s.accessCnt
-		c := s.batchC
-		s.batchC = make(chan bool)
-		s.batch = new(leveldb.Batch)
-		err := s.writeBatch(b, e, d, a)
-		// TODO: set this error on the batch, then tell the chunk
-		if err != nil {
-			log.Error(fmt.Sprintf("spawn batch write (%d entries): %v", b.Len(), err))
-		}
-		close(c)
-		for e > s.capacity {
-			s.collectGarbage(gcArrayFreeRatio)
-			e = s.entryCnt
+mainLoop:
+	for {
+		select {
+		case <-s.quit:
+			break mainLoop
+		case <-s.batchesC:
+			s.lock.Lock()
+			b := s.batch
+			e := s.entryCnt
+			d := s.dataIdx
+			a := s.accessCnt
+			c := s.batchC
+			s.batchC = make(chan bool)
+			s.batch = new(leveldb.Batch)
+			err := s.writeBatch(b, e, d, a)
+			// TODO: set this error on the batch, then tell the chunk
+			if err != nil {
+				log.Error(fmt.Sprintf("spawn batch write (%d entries): %v", b.Len(), err))
+			}
+			close(c)
+			for e > s.capacity {
+				// Collect garbage in a separate goroutine
+				// to be able to interrupt this loop by s.quit.
+				done := make(chan struct{})
+				go func() {
+					s.collectGarbage(gcArrayFreeRatio)
+					close(done)
+				}()
+
+				e = s.entryCnt
+				select {
+				case <-s.quit:
+					break mainLoop
```

Comment:
wouldn't this be more elegant to implement using `context.WithCancel`?

Reply:
I think that a single channel is the simplest for of cancelation in this case. We would have to check ctx.Done, the same way as it is done here with s.quit, but with checking ctx.Err and a bit more code to construct context and pass it around.